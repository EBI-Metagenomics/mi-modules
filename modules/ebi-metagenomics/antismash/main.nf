
process ANTISMASH {
    tag "$meta.id"
    label 'process_medium'

    // current antiSMASH bioconda container is buggy so we use our custom one (https://github.com/antismash/antismash/discussions/711)
    container "microbiome-informatics/antismash:7.1.0.1_2"

    input:
    tuple val(meta)         , path(contigs)
    tuple val(meta)         , path(genes)
    tuple val(meta)         , path(gbk_input)
    tuple path(antismash_db), val(db_version)

    output:
    tuple val(meta), path("${meta.id}_results/${meta.id}.gbk")  , emit: gbk
    tuple val(meta), path("${meta.id}_results/${meta.id}.json") , emit: json
    tuple val(meta), path("${meta.id}_antismash.tar.gz")        , emit: results_tarball
    path "versions.yml"                                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    if (gbk_input == []){
        """
        antismash \\
            -c ${task.cpus} \\
            ${args} \\
            --databases ${antismash_db} \\
            --genefinding-gff3 ${genes} \\
            --output-dir ${prefix}_results \\
            ${contigs}

        tar -czf ${prefix}_antismash.tar.gz ${prefix}_results

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            antiSMASH: \$(echo \$(antismash --version | sed 's/^antiSMASH //' ))
            antiSMASH database: ${db_version}
        END_VERSIONS
        """
    }
    else{
        """
        antismash \\
            -c ${task.cpus} \\
            ${args} \\
            --databases ${antismash_db} \\
            --output-dir ${prefix}_results \\
            ${gbk_input}

        tar -czf ${prefix}_antismash.tar.gz ${prefix}_results

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            antiSMASH: \$(echo \$(antismash --version | sed 's/^antiSMASH //' ))
            antiSMASH database: ${db_version}
        END_VERSIONS
        """
    }

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    touch ${prefix}_antismash.tar.gz
    mkdir ${prefix}_results/
    touch ${prefix}_results/${prefix}.gbk
    touch ${prefix}_results/${prefix}.json

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        antiSMASH: \$(echo \$(antismash --version | sed 's/^antiSMASH //' ))
        antiSMASH database: ${db_version}
    END_VERSIONS
    """
}
