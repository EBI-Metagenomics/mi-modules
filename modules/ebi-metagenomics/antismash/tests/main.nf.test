
nextflow_process {

    name "Test Process ANTISMASH"
    script "../main.nf"
    process "ANTISMASH"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "antismash"

    test("antismash") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file("${moduleDir}/tests/test.fasta", checkIfExists: true)
                    ]
                input[1] = [
                    [ id:'test', single_end:false ], // meta map
                    file("${moduleDir}/tests/test.gff", checkIfExists: true)
                    ]
                input[2] = [
                    file("${moduleDir}/tests/antismash_trimmed.tar.gz", checkIfExists: true),
                    "7.0.1"
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(path(process.out.gbk.get(0).get(1)).readLines()[0].contains("31398 bp")).match("gbk") },
                { assert snapshot(path(process.out.gbk.get(0).get(1)).readLines()[0].contains("NODE_1_length_31398_cov_44.290974")).match("json") }
            )
        }

    }

    test("antismash - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file("${moduleDir}/tests/test.fasta", checkIfExists: true)
                    ]
                input[1] = [
                    [ id:'test', single_end:false ], // meta map
                    file("${moduleDir}/tests/test.gff", checkIfExists: true)
                    ]
                input[2] = [
                    file("${moduleDir}/tests/antismash_trimmed.tar.gz", checkIfExists: true),
                    "7.0.1"
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.gbk).match("stub_gbk") }
            )
        }

    }

}
