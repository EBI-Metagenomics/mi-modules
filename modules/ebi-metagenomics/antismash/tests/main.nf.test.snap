{
    "gbk_input_json_out": {
        "content": [
            true
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-06-28T11:32:59.693222906"
    },
    "gff_input_gbk_out": {
        "content": [
            true
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-06-28T11:32:25.491985698"
    },
    "gbk_input_gbk_out": {
        "content": [
            true
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-06-28T11:32:59.676282068"
    },
    "gff_input_json_out": {
        "content": [
            true
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-06-28T11:32:25.521021989"
    },
    "stub_gbk": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.gbk:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-05-27T13:07:57.997518672"
    }
}